### PR TITLE
Update connection string builder to accept transparent gateway hostnames in simplified connection string

### PR DESCRIPTION
**Pull request by abhipsaMisra**
Currently we expect the hostname to be of the format "acme.azure-devices.net", in which case the IotHubName is "acme". For transparent gateway scenarios, we can simplify the input credentials to only specify the gateway device hostname, instead of having to specify both the IoT Hub hostname and the gateway device hostname.
In this case, the hostname will be of the format "myGatewayDevice", and will not have ".azure-devices.net" suffix.

For more details, see: https://docs.microsoft.com/en-us/azure/iot-edge/how-to-authenticate-downstream-device#retrieve-and-modify-connection-string

Fix for #1552

### CODE REVIEW

File: iothub/device/src/IotHubConnectionStringBuilder.cs [hunk 1]
```diff
@@ -294,9 +295,14 @@ private void SetIotHubName()
...
 
+            // We expect the hostname to be of the format "acme.azure-devices.net", in which case the IotHubName is "acme".
+            // For transparent gateway scenarios, we can simplify the input credentials to only specify the gateway device hostname,
+            // instead of having to specify both the IoT Hub hostname and the gateway device hostname.
+            // In this case, the hostname will be of the format "myGatewayDevice", and will not have ".azure-devices.net" suffix.
             if (IotHubName.IsNullOrWhiteSpace())
             {
-                throw new FormatException("Missing IOT hub name");
+                Logging.Info(this, $"Connecting to a gateway device with hostname=[{HostName}]");
+                IotHubName = HostName;
```
Comment (abhipsaMisra) [inline iothub/device/src/IotHubConnectionStringBuilder.cs:305]: There is no additional validation added here to test if this is indeed a gateway hostname or not since that is out of scope for the sdk.
In case an invalid host is specified here, service will throw the appropriate error.

File: iothub/device/src/IotHubConnectionStringBuilder.cs [hunk 2]
```diff
@@ -10,6 +10,7 @@
 using System.Net;
 using SharedAccessSignatureParser = Microsoft.Azure.Devices.Client.SharedAccessSignature;
 using System.Security.Cryptography.X509Certificates;
+using Microsoft.Azure.Devices.Shared;
```
Comment (drwill-ms) [inline iothub/device/src/IotHubConnectionStringBuilder.cs:13]: Is this sorted?

File: iothub/device/src/IotHubConnectionStringBuilder.cs [hunk 3]
```diff
@@ -294,9 +295,14 @@ private void SetIotHubName()
         {
             IotHubName = GetIotHubName(HostName);
 
+            // We expect the hostname to be of the format "acme.azure-devices.net", in which case the IotHubName is "acme".
+            // For transparent gateway scenarios, we can simplify the input credentials to only specify the gateway device hostname,
+            // instead of having to specify both the IoT Hub hostname and the gateway device hostname.
+            // In this case, the hostname will be of the format "myGatewayDevice", and will not have ".azure-devices.net" suffix.
             if (IotHubName.IsNullOrWhiteSpace())
```
Comment (drwill-ms) [inline iothub/device/src/IotHubConnectionStringBuilder.cs:302]: If it is null or whitespace or if it isn't??

File: iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs [hunk 4]
```diff
@@ -280,14 +316,13 @@ public void DeviceClientIotHubConnectionStringBuilderTest()
...
-            {
-                iotHubConnectionStringBuilder.HostName = "adshgfvyregferuehfiuehr";
-                Assert.Fail("Expected FormatException");
-            }
-            catch (FormatException)
-            {               
-            }
+            iotHubConnectionStringBuilder.HostName = "adshgfvyregferuehfiuehr";
+            iotHubConnectionStringBuilder.AuthenticationMethod = new DeviceAuthenticationWithRegistrySymmetricKey("Device1-.+%_#*?!(),=@;$'", "dGVzdFN0cmluZzE=");
+            Assert.IsTrue(iotHubConnectionStringBuilder.AuthenticationMethod is DeviceAuthenticationWithRegistrySymmetricKey);
```
Comment (drwill-ms) [inline iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs:321]: Should we switch to using fluent assertions?

File: iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs [hunk 5]
```diff
@@ -280,14 +316,13 @@ public void DeviceClientIotHubConnectionStringBuilderTest()
...
+            Assert.IsTrue(iotHubConnectionStringBuilder.HostName == "adshgfvyregferuehfiuehr");
+            Assert.IsTrue(iotHubConnectionStringBuilder.SharedAccessSignature == null);
```
Comment (drwill-ms) [inline iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs:323]: If not, Assert.IsNull

File: iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs [hunk 6]
```diff
@@ -280,14 +316,13 @@ public void DeviceClientIotHubConnectionStringBuilderTest()
...
-                iotHubConnectionStringBuilder.HostName = "adshgfvyregferuehfiuehr";
-                Assert.Fail("Expected FormatException");
-            }
-            catch (FormatException)
-            {               
-            }
+            iotHubConnectionStringBuilder.HostName = "adshgfvyregferuehfiuehr";
+            iotHubConnectionStringBuilder.AuthenticationMethod = new DeviceAuthenticationWithRegistrySymmetricKey("Device1-.+%_#*?!(),=@;$'", "dGVzdFN0cmluZzE=");
+            Assert.IsTrue(iotHubConnectionStringBuilder.AuthenticationMethod is DeviceAuthenticationWithRegistrySymmetricKey);
+            Assert.IsTrue(iotHubConnectionStringBuilder.HostName == "adshgfvyregferuehfiuehr");
```
Comment (drwill-ms) [inline iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs:322]: If not, Assert.AreEqual

File: iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs [hunk 7]
```diff
@@ -280,14 +316,13 @@ public void DeviceClientIotHubConnectionStringBuilderTest()
...
-            try
-            {
-                iotHubConnectionStringBuilder.HostName = "adshgfvyregferuehfiuehr";
-                Assert.Fail("Expected FormatException");
-            }
-            catch (FormatException)
-            {               
-            }
+            iotHubConnectionStringBuilder.HostName = "adshgfvyregferuehfiuehr";
+            iotHubConnectionStringBuilder.AuthenticationMethod = new DeviceAuthenticationWithRegistrySymmetricKey("Device1-.+%_#*?!(),=@;$'", "dGVzdFN0cmluZzE=");
```
Comment (drwill-ms) [inline iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs:320]: Best pull these out as local constants to use in validation (below)

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/src/IotHubConnectionStringBuilder.cs (hunk 2 above)
Comment (abhipsaMisra) [reply iothub/device/src/IotHubConnectionStringBuilder.cs:13]: might not be, i'll make sure that it is

File: iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs (hunk 4 above)
Comment (abhipsaMisra) [reply iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs:321]: sure, i'll do that throughout this file

File: iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs (hunk 5 above)
Comment (abhipsaMisra) [reply iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs:344]: updated to use fluent assertions

File: iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs (hunk 6 above)
Comment (abhipsaMisra) [reply iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs]: updated to use fluent assertions

File: iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs (hunk 7 above)
Comment (abhipsaMisra) [reply iothub/device/tests/ConnectionString/DeviceClientConnectionStringTests.cs]: done
